### ISSUE TITLE
learning rate scheduler

### ISSUE THREAD

**Issue by IamTao** [mlbench/mlbench#50]
current learning rate scheduler failed to fit the case when we resume the training. The `last_epoch` will always set to `-1` but it should be reset to the certain epoch number when we restart the training.

**Comment by IamTao** [comment]
[This line](https://github.com/mlbench/mlbench/blob/develop/mlbench/refimpls/pytorch/controlflow/controlflow.py#L130) also seems incorrect. Shouldn't it be `start_epoch = options.runtime['current_epoch'] + 1 if options.resume else 0`?

**Comment by IamTao** [comment]
Also, the modified learning rate scheduler only support to scale the learning rate based on the epoch, not the number of iterations. It might be a problem for large batch training for ImageNet when using many GPUs, e.g., linear warmup the learning rate from 0.1 to 0.1 * 128 during the first 5 epochs might be too sharp to hurt the performance.

**Comment by LiamHe** [comment]
@IamTao I was testing on the pull request branch. I did not found `last_epoch=-1` after resuming for both schedulers. The scheduler is saved and loaded by `scheduler.state_dict()` and `scheduler.load_state_dict(checkpoint['scheduler'])` in https://github.com/mlbench/mlbench/blob/develop/mlbench/refimpls/pytorch/utils/checkpoint.py`. The `last_epoch` attribute is supposed to be saved.

**Comment by IamTao** [comment]
ok. sorry I completely forgot to check your `checkpoint.py` file. It is fine.

**Comment by LiamHe** [comment]
In the `develop` branch, I use `LambdaLR` and `MultiStepLR ` in `torch.optim.lr_scheduler`. `torch.optim.lr_scheduler` uses `epoch` as the default interval between two `step`s, even though we can call `scheduelr.step()` in each epoch where the `epoch` is in fact a `batch`.
So I was not sure if we should keep their base class or not. In the pull request branch, I added one more argument to decide which interval to update. See [here](https://github.com/mlbench/mlbench/blob/refactor/controlflow/mlbench/refimpls/pytorch/controlflow/controlflow.py#L28) and [here](https://github.com/mlbench/mlbench/blob/refactor/controlflow/mlbench/refimpls/pytorch/controlflow/controlflow.py#L117).

**Comment by martinjaggi** [comment]
strictly speaking we don not urgently need to offer restarting ability for training. simplicity of training algorithm is preferred. nevertheless, we do need to support writing of checkpoints easily for sure. (and using those for evaluation, separately)

**Comment by LiamHe** [comment]
@martinjaggi The pytorch checkpoint methods are easy to use, all we need to do is to include all of the necessary modules. Two flags `resume` and `run_id` are used to determine the path to checkpoint. (We only assume users don't change the checkpoint directory when they resume an experiment).
The `save` and `resume` function in checkpoint module have the following signature:
```python
def save(options, model, optimizer, scheduler, is_best):
def resume(options, model, optimizer, scheduler):
```
where `options` contains the checkpoint path, and `options.runtime`, `model`, `optimizer`, `scheduler` are the stateful modules.

**Comment by LiamHe** [comment]
@IamTao Every time `scheduler.step()` is called, the `scheduler.epoch` do increase by 1. But the `scheduler.epoch` itself is not directly linked to the real training `epoch`; it is linked to the number of times `scheduler.step()` being called. 

If one decides the use `epoch` or `batch`, then they also need to provide a distinct argument or fix the behavior in `get_scheduler ` or a same argument with different values for two circumstances. It depends on how much freedom we want the users to have.

`train_epochs` and `max_train_steps` are kind of confusing. In mlperf, they use the first under common settings; the second one can be None, but when specified, it will be used to limit the training steps for performance check. It is not essential, we may decide to keep it or discard it later.
